### PR TITLE
feature/inference analysis dot

### PR DESCRIPTION
**Pull request by Superjomn**
This is the framework for inference analysis

This PR is a part of [the full code](https://github.com/PaddlePaddle/Paddle/pull/10459).

### CODE REVIEW

File: paddle/fluid/inference/analysis/helper.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+  iterator_range(const IteratorT &begin, const IteratorT &end)
+      : begin_(begin), end_(end) {}
+
+  const IteratorT &begin() const { return begin_; }
+  const IteratorT &end() const { return end_; }
+};
+
+/*
+ * An registry helper class, with its records keeps the order they registers.
```
Comment (luotao1) [inline paddle/fluid/inference/analysis/helper.h]: A registry helper class keeps the order they register?

File: python/paddle/fluid/metrics.py [hunk 2]
```diff
@@ -251,7 +251,7 @@ def update(self, distances, seq_num):
         self.instance_error += seq_num - seq_right_count
```
Comment (luotao1) [inline python/paddle/fluid/metrics.py]: 这个PR为什么需要修改metrics.py？

File: paddle/fluid/inference/analysis/node_tester.cc [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+#include "paddle/fluid/inference/analysis/node.h"
+
+#include <gtest/gtest.h>
+
+namespace paddle {
+namespace inference {
+namespace analysis {
+
+TEST(Node, Attr) {
```
Comment (luotao1) [inline paddle/fluid/inference/analysis/node_tester.cc]: node_tester.cc没有内容的话，可以先删掉。

File: paddle/fluid/inference/analysis/node.h [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+class NodeMap;
+
+/*
+ * Node Representation.
+ *
+ * This is a very important class for analysis. It is the base class of all
+ * nodes computed by a program that may be used as operands to other nodes.
+ * Node is the super class of other important classes such as Function and
+ * Value, some nodes can have a name.
```
Comment (luotao1) [inline paddle/fluid/inference/analysis/node.h]: some nodes can have a name：为什么有些可以有名字，有些没有呢？

Comment (Superjomn) [reply paddle/fluid/inference/analysis/node.h]: Value 有名字，Function没有； 类比 variable 有名字，OP没有名字

Comment (luotao1) [reply paddle/fluid/inference/analysis/node.h]: Get it 可以加下注释。

File: paddle/fluid/inference/analysis/node.h [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+ */
+class Node {
+ public:
+  // Node type. NOTE the new node types should add here.
```
Comment (luotao1) [inline paddle/fluid/inference/analysis/node.h]: NOTE that the new node types should be added here.

File: paddle/fluid/inference/analysis/node.h [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+  enum class Type { kNone = -1, kFunction, kValue, kFunctionBlock };
```
Comment (luotao1) [inline paddle/fluid/inference/analysis/node.h]: 1. kNone是用在什么地方呢？
2. kFunction->kOp?
3. kFunctionBlock->kBlock?

File: paddle/fluid/inference/analysis/node.h [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+  Node() = default;
+
+  // Cast to a subclass type, Function for example.
+  template <typename Subclass>
+  Subclass &As() {
+    return *reinterpret_cast<Subclass *>(this);
+  }
+
+  // Formatted representation of this Node.
```
Comment (luotao1) [inline paddle/fluid/inference/analysis/node.h]: Formatted->Format the

File: paddle/fluid/inference/analysis/node.h [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+  virtual std::string repr() const {
+    return name() + "(" + std::to_string(id()) + ")";
+  }
+
+  // DOT node representation. One Node type can customize its own node
+  // representation.
+  virtual std::vector<Dot::Attr> dot_attrs() const {
+    return std::vector<Dot::Attr>({Dot::Attr("style", "filled")});
```
Comment (luotao1) [inline paddle/fluid/inference/analysis/node.h]: 1.style filled是表示什么？
2.One Node type can customize its own node？是说每一种node type都有自定义？

File: paddle/fluid/inference/analysis/dot.cc [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/inference/analysis/dot.h"
+
+namespace paddle {
+namespace inference {
+namespace analysis {
+size_t Dot::counter = 0;
```
[the remark this answers is not in the document]
Comment (Superjomn) [reply paddle/fluid/inference/analysis/dot.cc]: reference https://stackoverflow.com/questions/18860895/how-to-initialize-static-members-in-the-header

the static member should be booted in a src.

File: paddle/fluid/inference/CMakeLists.txt [hunk 10]
```diff
@@ -21,6 +21,8 @@ endif()
 
 if(WITH_TESTING)
   add_subdirectory(tests/book)
+  # analysis test depends the models that generate by tests/book
```
[the remark this answers is not in the document]
Comment (Superjomn) [reply paddle/fluid/inference/CMakeLists.txt]: done

File: paddle/fluid/inference/CMakeLists.txt [hunk 11]
```diff
@@ -21,6 +21,8 @@ endif()
 
 if(WITH_TESTING)
   add_subdirectory(tests/book)
+  # analysis test depends the models that generate by python/paddle/fluid/tests/book
```
Comment (luotao1) [inline paddle/fluid/inference/CMakeLists.txt]: 24行放到23行前面
both tests/book and analysis depend on the models that generated by python/paddle/fluid/tests/book

File: paddle/fluid/inference/analysis/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1 @@
+cc_test(test_dot SRCS dot.cc dot.h dot_tester.cc)
```
Comment (luotao1) [inline paddle/fluid/inference/analysis/CMakeLists.txt]: cc_test(test_dot SRCS dot.cc dot_tester.cc)
不需要.h

File: paddle/fluid/inference/analysis/dot_tester.cc [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    dot->AddEdge("a", "c", {});
+  }
+
+  std::unique_ptr<Dot> dot;
+};
+
+TEST_F(DotTester, Build) {
+  auto codes = dot->Build();
+  // Output the DOT language code, the generated codes are too long to compare
+  // the string.
```
Comment (luotao1) [inline paddle/fluid/inference/analysis/dot_tester.cc]: 可以注释下结果，不需要比较。

**Pull request merged**